File: tests/upstream.test.ts
import { afterEach, beforeEach, describe, expect, test } from "bun:test";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { createHash } from "node:crypto";
import { decodeJwt } from "jose";
import type { Config } from "../src/config.ts";
import { DpopKeyManager, canonicalHtu } from "../src/dpop.ts";
import { UpstreamClient, pageInfo, type TokenProvider } from "../src/upstream.ts";
import { logger as baseLogger, type Logger } from "../src/logger.ts";

const ADAPTER = "https://adapter.example.com";
let home: string;

function cfg(overrides: Partial<Config> = {}): Config {
  return {
    ADAPTER_BASE_URL: ADAPTER,
    OKTA_CLIENT_ID: "cid",
    AGENT_ID: "agent-1",
    OKTA_REDIRECT_PORT: 0,
    OKTA_SCOPES: "openid offline_access",
    DPOP_ALG: "ES256",
    DPOP_KEY_MODE: "persistent",
    BRIDGE_HOME: home,
    HTTP_TIMEOUT_MS: 30000,
    LOG_LEVEL: "error",
    ...overrides,
  };
}

interface RecordedCall {
  url: string;
  init: RequestInit;
}

function mockFetch(handlers: Array<() => Response | Promise<Response>>): {
  fetch: typeof fetch;
  calls: RecordedCall[];
} {
  const calls: RecordedCall[] = [];
  const fn = (async (url: Parameters<typeof fetch>[0], init?: RequestInit) => {
    calls.push({ url: String(url), init: init ?? {} });
    const h = handlers[Math.min(calls.length - 1, handlers.length - 1)]!;
    return await h();
  }) as unknown as typeof fetch;
  return { fetch: fn, calls };
}

function jsonResponse(body: unknown, status = 200, headers: Record<string, string> = {}): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { "Content-Type": "application/json", ...headers },
  });
}

function headersOf(call: RecordedCall): Record<string, string> {
  return call.init.headers as Record<string, string>;
}

/** A TokenProvider stub that hands out a sequence of tokens. */
function stubTokens(...tokens: string[]): TokenProvider & { cleared: number; issued: string[] } {
  let i = 0;
  const issued: string[] = [];
  return {
    cleared: 0,
    issued,
    async getAccessToken() {
      const t = tokens[Math.min(i, tokens.length - 1)]!;
      i += 1;
      issued.push(t);
      return t;
    },
    clearStored() {
      this.cleared += 1;
    },
  };
}

const authorizeFn = async () => ({
  code: "c",
  redirectUri: "http://127.0.0.1:0/callback",
  verifier: "v",
});
const REQ = { jsonrpc: "2.0", id: 1, method: "tools/list", params: {} };

async function km(): Promise<DpopKeyManager> {
  return DpopKeyManager.create(cfg(), baseLogger);
}

beforeEach(() => {
  home = mkdtempSync(join(tmpdir(), "okta-mcp-bridge-up-"));
});
afterEach(() => {
  rmSync(home, { recursive: true, force: true });
});

describe("UpstreamClient.forward auth headers", () => {
  test("attaches Authorization: DPoP, X-MCP-Agent, and a DPoP proof", async () => {
    const { fetch, calls } = mockFetch([() => jsonResponse({ jsonrpc: "2.0", id: 1, result: {} })]);
    const client = new UpstreamClient(cfg(), await km(), stubTokens("tok-abc"), baseLogger, { fetch });
    await client.forward(REQ, authorizeFn);
    const h = headersOf(calls[0]!);
    expect(h.Authorization).toBe("DPoP tok-abc");
    expect(h["X-MCP-Agent"]).toBe("agent-1");
    expect(h.DPoP).toBeDefined();
  });

  test("the proof has htm POST, canonical htu '<base>/', and ath matching the token", async () => {
    const { fetch, calls } = mockFetch([() => jsonResponse({ jsonrpc: "2.0", id: 1, result: {} })]);
    const client = new UpstreamClient(cfg(), await km(), stubTokens("tok-xyz"), baseLogger, { fetch });
    await client.forward(REQ, authorizeFn);
    const proof = decodeJwt(headersOf(calls[0]!).DPoP!) as Record<string, unknown>;
    expect(proof.htm).toBe("POST");
    expect(proof.htu).toBe(canonicalHtu(`${ADAPTER}/`));
    const expectedAth = createHash("sha256").update("tok-xyz", "utf8").digest("base64url");
    expect(proof.ath).toBe(expectedAth);
  });

  test("forwardUnauthed sends NO Authorization/DPoP/X-MCP-Agent headers", async () => {
    const { fetch, calls } = mockFetch([() => jsonResponse({ jsonrpc: "2.0", id: 1, result: {} })]);
    const client = new UpstreamClient(cfg(), await km(), stubTokens("tok"), baseLogger, { fetch });
    await client.forwardUnauthed({ jsonrpc: "2.0", id: 1, method: "initialize" });
    const h = headersOf(calls[0]!);
    expect(h.Authorization).toBeUndefined();
    expect(h.DPoP).toBeUndefined();
    expect(h["X-MCP-Agent"]).toBeUndefined();
  });
});

describe("UpstreamClient sessions + SSE", () => {
  test("captures Mcp-Session-Id from a response and sends it next time", async () => {
    const { fetch, calls } = mockFetch([
      () => jsonResponse({ jsonrpc: "2.0", id: 1, result: {} }, 200, { "Mcp-Session-Id": "sess-9" }),
      () => jsonResponse({ jsonrpc: "2.0", id: 2, result: {} }),
    ]);
    const client = new UpstreamClient(cfg(), await km(), stubTokens("t"), baseLogger, { fetch });
    await client.forward(REQ, authorizeFn);
    await client.forward({ ...REQ, id: 2 }, authorizeFn);
    expect(headersOf(calls[0]!)["Mcp-Session-Id"]).toBeUndefined();
    expect(headersOf(calls[1]!)["Mcp-Session-Id"]).toBe("sess-9");
  });

  test("parses an SSE response body (last data: line)", async () => {
    const sse =
      "event: message\ndata: {\"jsonrpc\":\"2.0\",\"id\":1,\"result\":{\"tools\":[\"a\"]}}\n\n";
    const { fetch } = mockFetch([
      () => new Response(sse, { status: 200, headers: { "Content-Type": "text/event-stream" } }),
    ]);
    const client = new UpstreamClient(cfg(), await km(), stubTokens("t"), baseLogger, { fetch });
    const res = (await client.forward(REQ, authorizeFn)) as { result?: { tools?: string[] } };
    expect(res.result?.tools).toEqual(["a"]);
  });
});

describe("UpstreamClient 401 recovery", () => {
  test("401 use_dpop_nonce caches the nonce and retries once (second proof has nonce)", async () => {
    const { fetch, calls } = mockFetch([
      () =>
        jsonResponse({ error: "use_dpop_nonce" }, 401, {
          "WWW-Authenticate": 'DPoP error="use_dpop_nonce"',
          "DPoP-Nonce": "rs-nonce-1",
        }),
      () => jsonResponse({ jsonrpc: "2.0", id: 1, result: {} }),
    ]);
    const tokens = stubTokens("tok-1");
    const client = new UpstreamClient(cfg(), await km(), tokens, baseLogger, { fetch });
    await client.forward(REQ, authorizeFn);
    expect(calls.length).toBe(2);
    expect(tokens.cleared).toBe(0);
    const proof0 = decodeJwt(headersOf(calls[0]!).DPoP!) as Record<string, unknown>;
    const proof1 = decodeJwt(headersOf(calls[1]!).DPoP!) as Record<string, unknown>;
    expect(proof0.nonce).toBeUndefined();
    expect(proof1.nonce).toBe("rs-nonce-1");
  });

  test("401 without nonce clears the token, re-acquires, and retries once", async () => {
    const { fetch, calls } = mockFetch([
      () => jsonResponse({ jsonrpc: "2.0", id: 1, error: { code: -32000, message: "unauthorized" } }, 401),
      () => jsonResponse({ jsonrpc: "2.0", id: 1, result: {} }),
    ]);
    const tokens = stubTokens("tok-old", "tok-new");
    const client = new UpstreamClient(cfg(), await km(), tokens, baseLogger, { fetch });
    await client.forward(REQ, authorizeFn);
    expect(calls.length).toBe(2);
    expect(tokens.cleared).toBe(1);
    expect(headersOf(calls[0]!).Authorization).toBe("DPoP tok-old");
    expect(headersOf(calls[1]!).Authorization).toBe("DPoP tok-new");
  });

  test("persistent 401 yields a JSON-RPC error -32001 (no throw)", async () => {
    const { fetch } = mockFetch([() => jsonResponse({ error: "denied" }, 401)]);
    const client = new UpstreamClient(cfg(), await km(), stubTokens("t1", "t2"), baseLogger, { fetch });
    const res = (await client.forward(REQ, authorizeFn)) as { error?: { code?: number }; id?: unknown };
    expect(res.error?.code).toBe(-32001);
    expect(res.id).toBe(1);
  });

  test("a timeout yields a JSON-RPC error (no throw)", async () => {
    // Never resolves; aborts when the send AbortController fires.
    const hangingFetch = ((url: unknown, init?: RequestInit) =>
      new Promise((_resolve, reject) => {
        init?.signal?.addEventListener("abort", () =>
          reject(new DOMException("aborted", "AbortError")),
        );
      })) as unknown as typeof fetch;
    const client = new UpstreamClient(
      cfg({ HTTP_TIMEOUT_MS: 5 }),
      await km(),
      stubTokens("t"),
      baseLogger,
      { fetch: hangingFetch },
    );
    const res = (await client.forward(REQ, authorizeFn)) as { error?: { code?: number } };
    expect(res.error?.code).toBe(-32000);
  });

  test("forwardUnauthed maps a network error to a JSON-RPC error (no throw)", async () => {
    const failing = (() => Promise.reject(new Error("ECONNREFUSED"))) as unknown as typeof fetch;
    const client = new UpstreamClient(cfg(), await km(), stubTokens("t"), baseLogger, { fetch: failing });
    const res = (await client.forwardUnauthed({ jsonrpc: "2.0", id: 7, method: "initialize" })) as {
      error?: { code?: number };
      id?: unknown;
    };
    expect(res.error?.code).toBe(-32000);
    expect(res.id).toBe(7);
  });
});

/** A logger that records emitted records (event + fields) per level. */
function recordingLogger(): {
  logger: Logger;
  records: Array<{ level: string; event: string; fields?: Record<string, unknown> }>;
} {
  const records: Array<{ level: string; event: string; fields?: Record<string, unknown> }> = [];
  const make = (): Logger => ({
    log: (level, event, fields) => records.push({ level, event, fields }),
    debug: (event, fields) => records.push({ level: "debug", event, fields }),
    info: (event, fields) => records.push({ level: "info", event, fields }),
    warn: (event, fields) => records.push({ level: "warn", event, fields }),
    error: (event, fields) => records.push({ level: "error", event, fields }),
    withCorrelation: () => make(),
  });
  return { logger: make(), records };
}

function pageRecord(records: ReturnType<typeof recordingLogger>["records"]) {
  return records.find((r) => r.event === "mcp.response.page")?.fields;
}

describe("pageInfo (pagination diagnostic helper)", () => {
  test("reports nextCursor + item_count for a tools/list result", () => {
    const info = pageInfo("tools/list", {
      jsonrpc: "2.0",
      id: 1,
      result: { tools: [{ name: "a" }, { name: "b" }], nextCursor: "page2" },
    });
    expect(info).toEqual({ has_next_cursor: true, item_count: 2 });
  });

  test("has_next_cursor false when absent or empty; counts the right array", () => {
    expect(pageInfo("resources/list", { result: { resources: [1, 2, 3] } })).toEqual({
      has_next_cursor: false,
      item_count: 3,
    });
    expect(pageInfo("tools/list", { result: { tools: [], nextCursor: "" } })).toEqual({
      has_next_cursor: false,
      item_count: 0,
    });
  });

  test("returns null for non-list methods", () => {
    expect(pageInfo("tools/call", { result: { content: [] } })).toBeNull();
    expect(pageInfo("initialize", { result: {} })).toBeNull();
  });
});

describe("UpstreamClient pagination diagnostic log", () => {
  test("logs mcp.response.page with nextCursor + count for a JSON tools/list", async () => {
    const { logger, records } = recordingLogger();
    const { fetch } = mockFetch([
      () =>
        jsonResponse({
          jsonrpc: "2.0",
          id: 1,
          result: { tools: [{ name: "a" }, { name: "b" }, { name: "c" }], nextCursor: "next-25" },
        }),
    ]);
    const client = new UpstreamClient(cfg(), await km(), stubTokens("t"), logger, { fetch });
    await client.forward(REQ, authorizeFn);
    const page = pageRecord(records);
    expect(page).toBeDefined();
    expect(page?.has_next_cursor).toBe(true);
    expect(page?.item_count).toBe(3);
    expect(page?.content_type).toContain("application/json");
  });

  test("preserves nextCursor through an SSE tools/list (guards the last-data-line parse)", async () => {
    const { logger, records } = recordingLogger();
    const sse =
      "event: message\ndata: " +
      JSON.stringify({ jsonrpc: "2.0", id: 1, result: { tools: [{ name: "a" }], nextCursor: "sse-next" } }) +
      "\n\n";
    const { fetch } = mockFetch([
      () => new Response(sse, { status: 200, headers: { "Content-Type": "text/event-stream" } }),
    ]);
    const client = new UpstreamClient(cfg(), await km(), stubTokens("t"), logger, { fetch });
    await client.forward(REQ, authorizeFn);
    const page = pageRecord(records);
    expect(page?.has_next_cursor).toBe(true);
    expect(page?.item_count).toBe(1);
    expect(page?.content_type).toContain("text/event-stream");
  });

  test("does not log mcp.response.page for non-list methods", async () => {
    const { logger, records } = recordingLogger();
    const { fetch } = mockFetch([() => jsonResponse({ jsonrpc: "2.0", id: 1, result: { content: [] } })]);
    const client = new UpstreamClient(cfg(), await km(), stubTokens("t"), logger, { fetch });
    await client.forward({ jsonrpc: "2.0", id: 1, method: "tools/call", params: {} }, authorizeFn);
    expect(pageRecord(records)).toBeUndefined();
  });
});

File: internal/upstream/upstream_test.go
package upstream

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"io"
	"net/http"
	"strings"
	"testing"

	"github.com/joewitt99/bridge-mcp-client/internal/config"
	"github.com/joewitt99/bridge-mcp-client/internal/dpop"
	"github.com/joewitt99/bridge-mcp-client/internal/logx"
	"github.com/joewitt99/bridge-mcp-client/internal/oauth"
)

const adapter = "https://adapter.example.com"

var reqBytes = []byte(`{"jsonrpc":"2.0","id":1,"method":"tools/list","params":{}}`)

func authFn() (oauth.AuthCodeResult, error) { return oauth.AuthCodeResult{}, nil }

type doerFunc func(*http.Request) (*http.Response, error)

func (f doerFunc) Do(r *http.Request) (*http.Response, error) { return f(r) }

type recCall struct {
	headers http.Header
	body    []byte
}

func mockDoer(t *testing.T, resps ...*http.Response) (oauth.Doer, *[]recCall) {
	t.Helper()
	calls := &[]recCall{}
	i := 0
	d := doerFunc(func(r *http.Request) (*http.Response, error) {
		b, _ := io.ReadAll(r.Body)
		*calls = append(*calls, recCall{headers: r.Header.Clone(), body: b})
		if i >= len(resps) {
			t.Fatalf("unexpected upstream request #%d", i+1)
		}
		resp := resps[i]
		i++
		return resp, nil
	})
	return d, calls
}

func resp(status int, body any, headers map[string]string) *http.Response {
	var bb []byte
	if s, ok := body.(string); ok {
		bb = []byte(s)
	} else {
		bb, _ = json.Marshal(body)
	}
	h := http.Header{}
	h.Set("Content-Type", "application/json")
	for k, v := range headers {
		h.Set(k, v)
	}
	return &http.Response{StatusCode: status, Body: io.NopCloser(bytes.NewReader(bb)), Header: h}
}

func tcfg(t *testing.T, overrides map[string]string) (config.Config, *dpop.KeyManager) {
	t.Helper()
	env := map[string]string{
		"ADAPTER_BASE_URL": adapter,
		"OKTA_CLIENT_ID":   "cid",
		"AGENT_ID":         "agent-1",
		"BRIDGE_HOME":      t.TempDir(),
		"LOG_LEVEL":        "error",
	}
	for k, v := range overrides {
		env[k] = v
	}
	cfg, err := config.Load(env)
	if err != nil {
		t.Fatal(err)
	}
	km, err := dpop.NewKeyManager(cfg, nil)
	if err != nil {
		t.Fatal(err)
	}
	return cfg, km
}

type stubTokens struct {
	toks    []string
	i       int
	cleared int
}

func (s *stubTokens) GetAccessToken(_ context.Context, _ oauth.AuthorizeFn) (string, error) {
	tok := s.toks[min(s.i, len(s.toks)-1)]
	s.i++
	return tok, nil
}
func (s *stubTokens) ClearStored() error { s.cleared++; return nil }

func proofOf(t *testing.T, c recCall) map[string]any {
	t.Helper()
	parts := strings.Split(c.headers.Get("DPoP"), ".")
	if len(parts) != 3 {
		t.Fatalf("bad DPoP header")
	}
	b, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		t.Fatal(err)
	}
	var m map[string]any
	json.Unmarshal(b, &m)
	return m
}

func norm(m map[string]any) map[string]any {
	b, _ := json.Marshal(m)
	var out map[string]any
	json.Unmarshal(b, &out)
	return out
}

func TestForwardAuthHeaders(t *testing.T) {
	cfg, km := tcfg(t, nil)
	doer, calls := mockDoer(t, resp(200, map[string]any{"jsonrpc": "2.0", "id": 1, "result": map[string]any{}}, nil))
	c := New(cfg, km, &stubTokens{toks: []string{"tok-abc"}}, nil, Deps{Doer: doer})
	c.Forward(context.Background(), reqBytes, authFn)
	h := (*calls)[0].headers
	if h.Get("Authorization") != "DPoP tok-abc" || h.Get("X-MCP-Agent") != "agent-1" || h.Get("DPoP") == "" {
		t.Fatalf("bad auth headers: %v", h)
	}
}

func TestForwardProofHtmHtuAth(t *testing.T) {
	cfg, km := tcfg(t, nil)
	doer, calls := mockDoer(t, resp(200, map[string]any{"jsonrpc": "2.0", "id": 1, "result": map[string]any{}}, nil))
	c := New(cfg, km, &stubTokens{toks: []string{"tok-xyz"}}, nil, Deps{Doer: doer})
	c.Forward(context.Background(), reqBytes, authFn)
	pl := proofOf(t, (*calls)[0])
	wantHTU, _ := dpop.CanonicalHTU(adapter + "/")
	sum := sha256.Sum256([]byte("tok-xyz"))
	wantAth := base64.RawURLEncoding.EncodeToString(sum[:])
	if pl["htm"] != "POST" || pl["htu"] != wantHTU || pl["ath"] != wantAth {
		t.Fatalf("proof = %v; want htm POST htu %s ath %s", pl, wantHTU, wantAth)
	}
}

func TestForwardUnauthedNoAuthHeaders(t *testing.T) {
	cfg, km := tcfg(t, nil)
	doer, calls := mockDoer(t, resp(200, map[string]any{"jsonrpc": "2.0", "id": 1, "result": map[string]any{}}, nil))
	c := New(cfg, km, &stubTokens{toks: []string{"t"}}, nil, Deps{Doer: doer})
	c.ForwardUnauthed(context.Background(), []byte(`{"jsonrpc":"2.0","id":1,"method":"initialize"}`))
	h := (*calls)[0].headers
	if h.Get("Authorization") != "" || h.Get("DPoP") != "" || h.Get("X-MCP-Agent") != "" {
		t.Fatalf("unauthed request leaked auth headers: %v", h)
	}
}

func TestSessionCaptureAndReuse(t *testing.T) {
	cfg, km := tcfg(t, nil)
	doer, calls := mockDoer(t,
		resp(200, map[string]any{"jsonrpc": "2.0", "id": 1, "result": map[string]any{}}, map[string]string{"Mcp-Session-Id": "sess-9"}),
		resp(200, map[string]any{"jsonrpc": "2.0", "id": 2, "result": map[string]any{}}, nil),
	)
	c := New(cfg, km, &stubTokens{toks: []string{"t"}}, nil, Deps{Doer: doer})
	c.Forward(context.Background(), reqBytes, authFn)
	c.Forward(context.Background(), reqBytes, authFn)
	if (*calls)[0].headers.Get("Mcp-Session-Id") != "" {
		t.Error("first request should not carry a session id")
	}
	if (*calls)[1].headers.Get("Mcp-Session-Id") != "sess-9" {
		t.Error("second request should reuse the captured session id")
	}
}

func TestSSEParsed(t *testing.T) {
	cfg, km := tcfg(t, nil)
	sse := "event: message\ndata: {\"jsonrpc\":\"2.0\",\"id\":1,\"result\":{\"tools\":[\"a\"]}}\n\n"
	doer, _ := mockDoer(t, resp(200, sse, map[string]string{"Content-Type": "text/event-stream"}))
	c := New(cfg, km, &stubTokens{toks: []string{"t"}}, nil, Deps{Doer: doer})
	out := norm(c.Forward(context.Background(), reqBytes, authFn))
	result := out["result"].(map[string]any)
	tools := result["tools"].([]any)
	if len(tools) != 1 || tools[0] != "a" {
		t.Fatalf("SSE body not parsed: %v", out)
	}
}

func TestNonceRetry(t *testing.T) {
	cfg, km := tcfg(t, nil)
	doer, calls := mockDoer(t,
		resp(401, map[string]any{"error": "use_dpop_nonce"}, map[string]string{
			"WWW-Authenticate": `DPoP error="use_dpop_nonce"`, "DPoP-Nonce": "rs-nonce-1",
		}),
		resp(200, map[string]any{"jsonrpc": "2.0", "id": 1, "result": map[string]any{}}, nil),
	)
	tokens := &stubTokens{toks: []string{"tok-1"}}
	c := New(cfg, km, tokens, nil, Deps{Doer: doer})
	c.Forward(context.Background(), reqBytes, authFn)
	if len(*calls) != 2 || tokens.cleared != 0 {
		t.Fatalf("calls=%d cleared=%d", len(*calls), tokens.cleared)
	}
	if _, ok := proofOf(t, (*calls)[0])["nonce"]; ok {
		t.Error("first proof should have no nonce")
	}
	if proofOf(t, (*calls)[1])["nonce"] != "rs-nonce-1" {
		t.Error("retry proof should carry the resource nonce")
	}
}

func TestStaleTokenRetry(t *testing.T) {
	cfg, km := tcfg(t, nil)
	doer, calls := mockDoer(t,
		resp(401, map[string]any{"jsonrpc": "2.0", "id": 1, "error": map[string]any{"code": -32000, "message": "unauthorized"}}, nil),
		resp(200, map[string]any{"jsonrpc": "2.0", "id": 1, "result": map[string]any{}}, nil),
	)
	tokens := &stubTokens{toks: []string{"tok-old", "tok-new"}}
	c := New(cfg, km, tokens, nil, Deps{Doer: doer})
	c.Forward(context.Background(), reqBytes, authFn)
	if len(*calls) != 2 || tokens.cleared != 1 {
		t.Fatalf("calls=%d cleared=%d", len(*calls), tokens.cleared)
	}
	if (*calls)[0].headers.Get("Authorization") != "DPoP tok-old" || (*calls)[1].headers.Get("Authorization") != "DPoP tok-new" {
		t.Fatal("token not re-acquired on retry")
	}
}

func TestPersistent401(t *testing.T) {
	cfg, km := tcfg(t, nil)
	doer, _ := mockDoer(t,
		resp(401, map[string]any{"error": "denied"}, nil),
		resp(401, map[string]any{"error": "denied"}, nil),
	)
	c := New(cfg, km, &stubTokens{toks: []string{"t1", "t2"}}, nil, Deps{Doer: doer})
	out := norm(c.Forward(context.Background(), reqBytes, authFn))
	if out["error"].(map[string]any)["code"] != float64(-32001) || out["id"] != float64(1) {
		t.Fatalf("expected -32001 with id 1, got %v", out)
	}
}

func TestTimeout(t *testing.T) {
	cfg, km := tcfg(t, map[string]string{"HTTP_TIMEOUT_MS": "20"})
	hanging := doerFunc(func(r *http.Request) (*http.Response, error) {
		<-r.Context().Done()
		return nil, r.Context().Err()
	})
	c := New(cfg, km, &stubTokens{toks: []string{"t"}}, nil, Deps{Doer: hanging})
	out := norm(c.Forward(context.Background(), reqBytes, authFn))
	if out["error"].(map[string]any)["code"] != float64(-32000) {
		t.Fatalf("expected -32000 on timeout, got %v", out)
	}
}

func TestForwardUnauthedNetworkError(t *testing.T) {
	cfg, km := tcfg(t, nil)
	failing := doerFunc(func(*http.Request) (*http.Response, error) {
		return nil, io.ErrUnexpectedEOF
	})
	c := New(cfg, km, &stubTokens{toks: []string{"t"}}, nil, Deps{Doer: failing})
	out := norm(c.ForwardUnauthed(context.Background(), []byte(`{"jsonrpc":"2.0","id":7,"method":"initialize"}`)))
	if out["error"].(map[string]any)["code"] != float64(-32000) || out["id"] != float64(7) {
		t.Fatalf("expected -32000 with id 7, got %v", out)
	}
}

func Test404SessionRecovery(t *testing.T) {
	cfg, km := tcfg(t, nil)
	doer, calls := mockDoer(t,
		resp(200, map[string]any{"jsonrpc": "2.0", "id": 1, "result": map[string]any{}}, map[string]string{"Mcp-Session-Id": "sess-1"}),
		resp(404, map[string]any{"error": "session not found"}, nil),
		resp(200, map[string]any{"jsonrpc": "2.0", "id": 2, "result": map[string]any{"ok": true}}, nil),
	)
	c := New(cfg, km, &stubTokens{toks: []string{"t"}}, nil, Deps{Doer: doer})
	c.Forward(context.Background(), reqBytes, authFn) // establishes session
	out := norm(c.Forward(context.Background(), reqBytes, authFn))

	if len(*calls) != 3 {
		t.Fatalf("expected 3 requests, got %d", len(*calls))
	}
	if (*calls)[1].headers.Get("Mcp-Session-Id") != "sess-1" {
		t.Error("404 request should have carried the stale session")
	}
	if (*calls)[2].headers.Get("Mcp-Session-Id") != "" {
		t.Error("retry after 404 should drop the cleared session")
	}
	if out["result"].(map[string]any)["ok"] != true {
		t.Fatalf("expected recovered result, got %v", out)
	}
}

func TestPageInfoUnit(t *testing.T) {
	if p := pageInfo("tools/list", map[string]any{"result": map[string]any{
		"tools": []any{1, 2, 3}, "nextCursor": "p2"}}); p == nil || !p.hasNextCursor || p.itemCount != 3 {
		t.Fatalf("tools/list with cursor: %+v", p)
	}
	if p := pageInfo("resources/list", map[string]any{"result": map[string]any{"resources": []any{1}}}); p == nil || p.hasNextCursor || p.itemCount != 1 {
		t.Fatalf("resources/list no cursor: %+v", p)
	}
	if p := pageInfo("tools/call", map[string]any{"result": map[string]any{}}); p != nil {
		t.Fatalf("non-list should be nil: %+v", p)
	}
}

func TestPaginationDiagnosticLog(t *testing.T) {
	cfg, km := tcfg(t, nil)
	doer, _ := mockDoer(t, resp(200, map[string]any{
		"jsonrpc": "2.0", "id": 1,
		"result": map[string]any{"tools": []any{map[string]any{"name": "a"}, map[string]any{"name": "b"}}, "nextCursor": "next-25"},
	}, nil))
	buf := &bytes.Buffer{}
	c := New(cfg, km, &stubTokens{toks: []string{"t"}}, logx.NewWith(buf, "debug"), Deps{Doer: doer})
	c.Forward(context.Background(), reqBytes, authFn)
	s := buf.String()
	if !strings.Contains(s, `"event":"mcp.response.page"`) || !strings.Contains(s, `"has_next_cursor":true`) || !strings.Contains(s, `"item_count":2`) {
		t.Fatalf("missing/incorrect pagination diagnostic: %q", s)
	}
}
